### PR TITLE
[RFDAP-8109] Update version of sqlalchemy

### PR DESCRIPTION
**Pull request by ricardobz**
### Please check if the pull request fulfills these requirements:

- [x] Title is in the format `[RFDAP-XXXX] Short description of what this PR does`
  - Do not use the automatic suggestion obtained from the branch name.
- [x] All builds have passed
- [x] Tests for the changes have been added
- [x] The commit message includes the issue id `[RFDAP-XXXX]`
- [x] Squash, re-order, or reword commits to obtain a readable history for future reference
- [x] Run Claude Code review skill (publish and resolve comments)

### What's the purpose of this PR?

The flask-restalchemy now supports SQLAlchemy 2.x alongside the existing 1.4 support.

### Explain what you've done.

Replaced all legacy point-lookup patterns (Model.query.get(), session.query(Model).get()) with session.get(Model, id), which is the unified API accepted by both SQLAlchemy 1.4 and 2.x. Collection queries (session.query() → select()) were not migrated: doing so requires a coordinated change across resources.py, querybuilder.py, and the public query_modifier callable contract (a breaking change left for a follow-up).
Fixed get_db_session() to work with both Flask-SQLAlchemy 2.x and Flask-SQLAlchemy 3.x.
Replaced raw integer HTTP status codes with http.HTTPStatus enum throughout resources.py.
Migrated project metadata from setup.cfg/setup.py to pyproject.toml and switched dependency management to uv! 
Updated tox to test against both sqla14 and sqla20 environments.
Updated the deploy CI step to use Python 3.12.
Fixed minor bugs and typos.
Added some more tests to cover the functions workflow.
Removed empty test files!

### Is there any related pull request?

N/A

### How to test this feature/fix?

Clone this repo chekout this branch, follow the steps on readme to create the env and activate it (uv)!
To run the tests just use this cmd: `pytest src/flask_restalchemy/tests/`

How to test within rfdap:
After cloning the repo in /Projects and checking out this branch
`pixi run pip install -e ../flask-restalchemy`
Check if the installation succeeded (it should show the repo folder, not the installed env lib (rfdap/.pixi/envs/...)):
`pixi run python -c "import flask_restalchemy; print(flask_restalchemy.__file__)"`

To return to published lib version in .venv (pyproject.toml):
`pixi run pip uninstall flask-restalchemy`
`pixi run pip install flask-restalchemy`

### Additional information (screenshots, references, links):
SQLAlchemy 2.0 migration guide: https://docs.sqlalchemy.org/en/20/changelog/migration_20.html

[RFDAP-8109]

[RFDAP-8109]: https://esss.atlassian.net/browse/RFDAP-8109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/flask_restalchemy/tests/test_api_relations.py [hunk 1]
```diff
@@ -146,32 +148,44 @@ def test_property_pagination(client):
...
+
 def test_delete_on_relation_with_secondary(client):
-    jim = Employee.query.get(9)
+    jim = db.session.get(Employee, 9)
     assert jim is not None
     dep = jim.departments[0]
 
-    sarah = Employee.query.get(3)
+    sarah = db.session.get(Employee, 3)
     assert jim is not None
```
Comment (ricardobz) [inline src/flask_restalchemy/tests/test_api_relations.py]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**Bug: wrong variable asserted — copy-paste error**

```python
sarah = db.session.get(Employee, 3)
assert jim is not None   # ← should be `assert sarah is not None`
assert dep not in sarah.departments
```

`jim` was already guarded two lines above. This assertion is meant to guard `sarah`, but checks the wrong variable. If `session.get(Employee, 3)` ever returns `None`, the test will raise an `AttributeError` on `sarah.departments` instead of failing the assertion cleanly.

File: src/flask_restalchemy/tests/test_api_processors.py [hunk 2]
```diff
@@ -118,7 +119,7 @@ def test_delete_preprocessors(sample_api, client, mocker):
     )
 
     resp = client.delete("/employee/1")
-    assert resp.status_code == 204
+    assert resp.status_code == HTTPStatus.NO_CONTENT
     assert pre_mock.call_args == call(id=1)
     assert post_mock.call_args == call(("", 204), id=1)
```
Comment (ricardobz) [inline src/flask_restalchemy/tests/test_api_processors.py]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**Incomplete migration: raw `204` not converted to `HTTPStatus`**

```python
assert post_mock.call_args == call(("", 204), id=1)
```

Every other status code in this file was migrated to `HTTPStatus`, but this one was missed. The view now returns `("", HTTPStatus.NO_CONTENT)`, so the captured tuple contains `HTTPStatus.NO_CONTENT`, not `204`. The assertion passes today only because `HTTPStatus` inherits from `int`, but it is misleading and inconsistent with the rest of the file.

Should be:
```python
assert post_mock.call_args == call(("", HTTPStatus.NO_CONTENT), id=1)
```

File: src/flask_restalchemy/resources/resources.py [hunk 3]
```diff
@@ -123,12 +125,12 @@ def _db_session(self):
...
         if id is not None:
-            model = self._resource_model.query.get(id)
+            model = self._db_session.get(self._resource_model, id)
             if model is None:
-                return NOT_FOUND_ERROR, 404
+                return NOT_FOUND_ERROR, HTTPStatus.NOT_FOUND
             return self._serializer.dump(model)
         else:
-            query = self._resource_model.query
+            query = self._db_session.query(self._resource_model)
```
Comment (ricardobz) [inline src/flask_restalchemy/resources/resources.py:133]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**`session.query()` is deprecated in SQLAlchemy 2.0 — missing TODO with Jira ID**

```python
query = self._db_session.query(self._resource_model)
```

The PR description acknowledges that migrating collection queries is deferred because it would be a breaking change (it requires coordinating changes to `querybuilder.py` and the public `query_modifier` contract). That is a legitimate call, but two problems remain:

1. `session.query()` emits `LegacyAPIWarning` under SQLAlchemy 2.x, and `pyproject.toml` promotes that warning to an error. This line will cause `sqla20` test runs to fail unless something suppresses it — please verify CI is actually green for the `sqla20` tox environment.

2. Per team convention, TODOs must include a Jira task ID so the deferral is traceable. There should be a comment here like:
   ```python
   # TODO(RFDAP-XXXX): migrate to select() once querybuilder.py and the
   # query_modifier public API are updated.
   query = self._db_session.query(self._resource_model)
   ```

File: src/flask_restalchemy/tests/test_swagger_spec_gen.py [hunk 4]
```diff
@@ -1,44 +0,0 @@
-import pytest
```
Comment (ricardobz) [inline src/flask_restalchemy/tests/test_swagger_spec_gen.py:1]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**Deleting xfail tests removes intent without leaving any traceability**

All 44 lines in this file carried `@pytest.mark.xfail`, which signals "this feature exists and we intend to fix it eventually." Deleting the file silently drops that intent — no one reading the codebase will know swagger spec generation was ever a goal.

If the decision is that swagger spec generation is no longer supported, that should be made explicit: either via a note in the PR description pointing to a Jira ticket that formally deprecates the feature, or by replacing the file with a `pytest.skip` call and a tracking comment. Removing xfail tests is functionally a no-op for the test run, but the signal they carried is lost.

File: setup.py [hunk 5]
```diff
@@ -1,44 +1,3 @@
...
-        "Intended Audience :: Developers",
-        "License :: OSI Approved :: MIT License",
-        "Operating System :: OS Independent",
-        "Programming Language :: Python :: 3.6",
-        "Programming Language :: Python :: 3.7",
-        "Topic :: Database :: Front-Ends",
-        "Topic :: Internet :: WWW/HTTP",
-    ],
-)
+setup()
```
Comment (ricardobz) [inline setup.py]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**`setup.py` should be fully deleted**

```python
from setuptools import setup

setup()
```

All metadata has been moved to `pyproject.toml`, and `[build-system]` already declares `setuptools.build_meta`. An argument-less `setup()` does nothing except create confusion about where the source of truth is. Modern `setuptools` (≥61) does not require `setup.py` at all when `pyproject.toml` is present. This file should be removed.

File: src/flask_restalchemy/resources/resources.py [hunk 6]
```diff
@@ -208,27 +210,26 @@ def get(self, relation_id, id=None):
...
             if related_obj is None:
-                return NOT_FOUND_ERROR, 404
+                return NOT_FOUND_ERROR, HTTPStatus.NOT_FOUND
 
             # TODO: Is there a more efficient way than using getattr?
             relation_list_or_query = getattr(related_obj, self._relation_property.key)
             if isinstance(relation_list_or_query, InstrumentedList) or not hasattr(
                 relation_list_or_query, "paginate"
             ):
                 warnings.warn(
```
Comment (ricardobz) [inline src/flask_restalchemy/resources/resources.py:231]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**`warnings.warn()` missing `stacklevel` — warning points into library internals**

```python
warnings.warn(
    "Warning: relationship does not support pagination nor filter."
    'Use flask-sqlalchemy relationship with lazy="dynamic".'
)
```

Without `stacklevel`, Python reports this warning as originating from inside `resources.py`, not from the user's `add_relation()` call. The user sees a traceback into library internals rather than the line in their own code that caused the problem.

Same issue applies at the equivalent call in `CollectionPropertyResource.get()`.

Fix:
```python
warnings.warn(
    "Warning: relationship does not support pagination nor filter."
    'Use flask-sqlalchemy relationship with lazy="dynamic".',
    stacklevel=2,
)
```

File: src/flask_restalchemy/tests/test_query_callback.py [hunk 7]
```diff
@@ -132,7 +131,7 @@ def sample_property_query(parent_query, model):
     if parent_query:
```
Comment (ricardobz) [inline src/flask_restalchemy/tests/test_query_callback.py:131]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**Silent behavioral bug fix buried in migration noise — should be called out explicitly**

```python
# Before (wrong):
query = model.query.filter_by(model.id != self_id)

# After (correct):
query = db.session.query(model).filter(model.id != self_id)
```

`filter_by()` accepts keyword arguments for column equality — passing an SQLAlchemy binary expression (`model.id != self_id`) to it is a misuse. The expression is evaluated eagerly to a Python boolean before being passed, so the old code was effectively filtering on `True`/`False`, not on the column condition, producing silently wrong query results.

The fix is correct, but this is a real behavioral bug, not just a migration cleanup. It deserves a mention in the PR description so reviewers and future git-blame readers know the semantic change was intentional.

File: pyproject.toml [hunk 8]
```diff
@@ -1,14 +1,66 @@
...
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "License :: OSI Approved :: MIT License",
+    "Topic :: Database :: Front-Ends",
+    "Topic :: Internet :: WWW/HTTP",
+]
+dependencies = [
+    "flask>=1.0.0",
+    "flask-sqlalchemy>=2.3.0",
```
Comment (ricardobz) [inline pyproject.toml]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**Dependency constraint allows an incompatible combination of flask-sqlalchemy and sqlalchemy**

```toml
"flask-sqlalchemy>=2.3.0",
"sqlalchemy>=1.4,<3",
```

Flask-SQLAlchemy 2.x does **not** support SQLAlchemy 2.x — it declares a `sqlalchemy<2` upper bound. A user who resolves `flask-sqlalchemy==2.5.1` alongside `sqlalchemy==2.0.x` gets a broken environment with no pip-level error.

Flask-SQLAlchemy 3.0 supports both SQLAlchemy >=1.4 and 2.x, so bumping the lower bound to `>=3.0` covers both test scenarios without restricting the 1.4 path:

```toml
"flask-sqlalchemy>=3.0",
"sqlalchemy>=1.4,<3",
```

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -52,7 +52,7 @@ jobs:
     - name: Set up Python
       uses: actions/setup-python@v2
       with:
-        python-version: "3.6"
+        python-version: "3.12"
```
Comment (ricardobz) [inline .github/workflows/main.yml:55]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**`actions/setup-python@v2` is outdated (two lines above)**

The deploy job still uses `actions/setup-python@v2` (line 53) while the build job was updated to `actions/setup-python@v4`. `v2` predates built-in caching, free-threaded Python support, and several bug fixes. Please update the deploy job to match:

```yaml
uses: actions/setup-python@v4  # or v5
```

File: README.md [hunk 10]
```diff
@@ -43,6 +43,28 @@ api = Api(flask_app)
 api.add_model(User, "/user", serializer_class=UserSerializer)
 ```
 
+### Development Setup
```
Comment (ricardobz) [inline README.md:46]: > 🤖 *Comment generated by Claude Code (AI-assisted review)*

**Development setup instructions duplicated between `README.md` and `CLAUDE.md`**

The identical `uv python pin / uv sync / source .venv/activate` block appears verbatim in both files. When one is updated the other will silently drift.

Pick one as the authoritative source (typically `README.md` for end-users) and have the other reference it, or factor the instructions out so they live in a single place.

File: src/flask_restalchemy/resources/resources.py (hunk 3 above)
Comment (ricardobz) [reply src/flask_restalchemy/resources/resources.py:133]: Changed:
`model = self._resource_model.query.get(id)`
To:
`model = self._db_session.get(self._resource_model, id)`
Tested and it does not emit LegacyAPIWarning! 🤔

File: src/flask_restalchemy/tests/test_swagger_spec_gen.py (hunk 4 above)
Comment (ricardobz) [reply src/flask_restalchemy/tests/test_swagger_spec_gen.py:1]: I removed this test because it was doing nothing, and had missing functions in import... should I maintain?
Do we plan to get this to work properly? 🤷‍♂️

File: setup.py (hunk 5 above)
Comment (ricardobz) [reply setup.py]: Removed, also the references!

File: src/flask_restalchemy/resources/resources.py (hunk 6 above)
Comment (ricardobz) [reply src/flask_restalchemy/resources/resources.py:231]: I'm gonna leave as it is for now!

**Comment by ricardobz** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=ESSS_flask-restalchemy&pullRequest=71) **Quality Gate failed**
> Failed conditions  [2 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=ESSS_flask-restalchemy&pullRequest=71&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=ESSS_flask-restalchemy&pullRequest=71)

I'm not sure about that... we usually use the version itself!
{C51C2C7C-86E6-4E54-8960-C8EAB6367BA8}

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -18,8 +18,8 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: ["3.6", "3.7", "3.8", "3.9", "3.10"]
-        os: [ubuntu-20.04, windows-latest]
+        python: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
Comment (ricardobz) [inline .github/workflows/main.yml]: I removed 3.6 and 3.7 support due to problems resolving the env dependencies!
I'm considering removing these two ("3.8" and "3.9) too... Since I got these errors... what do you guys think!?
https://github.com/ESSS/flask-restalchemy/actions/runs/25872898407

Image

Tested suggestion by claude, but didn't work:
```
filterwarnings = [
    # serialchemy still uses Query.get() internally; suppress until that dependency is updated
    "ignore:.*Query.get.*:sqlalchemy.exc.LegacyAPIWarning",
```

File: src/flask_restalchemy/tests/test_swagger_spec_gen.py (hunk 4 above)
Comment (nobreconfrade) [reply src/flask_restalchemy/tests/test_swagger_spec_gen.py:1]: It seems the right call to remove it

File: .github/workflows/main.yml (hunk 11 above)
Comment (nobreconfrade) [reply .github/workflows/main.yml]: Do we even have apps running with < python 3.10?

**Review by nobreconfrade — APPROVED** [review]
Image

💯

File: src/flask_restalchemy/resources/resources.py (hunk 3 above)
Comment (lvoliveira) [reply src/flask_restalchemy/resources/resources.py:133]: The Query is actually not deprecated, although they are moving to a new standard with session and select. See https://docs.sqlalchemy.org/en/21/changelog/migration_20.html#migration-orm-usage.

File: src/flask_restalchemy/tests/test_swagger_spec_gen.py (hunk 4 above)
Comment (lvoliveira) [reply src/flask_restalchemy/tests/test_swagger_spec_gen.py:1]: remove it 👍

File: .github/workflows/main.yml (hunk 11 above)
Comment (lvoliveira) [reply .github/workflows/main.yml]: Since they have already reached end-of-life we can remove them.

image

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -0,0 +1,31 @@
+### Please check if the pull request fulfills these requirements:
+
+- [ ] Title is in the format `[RFDAP-XXXX] Short description of what this PR does`
```
Comment (lvoliveira) [inline .github/PULL_REQUEST_TEMPLATE.md]: This is a public repo, we don't actually enforce the need for a jira issue key.
```suggestion
- [ ] Title is in the format `Short description of what this PR does`
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  - Do not use the automatic suggestion obtained from the branch name.
+- [ ] All builds have passed
+- [ ] Tests for the changes have been added
+- [ ] The commit message includes the issue id `[RFDAP-XXXX]`
```
Comment (lvoliveira) [inline .github/PULL_REQUEST_TEMPLATE.md]: Same reason here.
```suggestion
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+- [ ] Squash, re-order, or reword commits to obtain a readable history for future reference
+- [ ] Run Claude Code review skill (publish and resolve comments)
```
Comment (lvoliveira) [inline .github/PULL_REQUEST_TEMPLATE.md]: This should not be required either. 
```suggestion
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+### How to test this feature/fix?
+
+<!--- Steps to verify the change works, including any edge cases worth checking -->
+
+### Additional information (screenshots, references, links):
+
+<!--- Replace XXXX with the issue code -->
+
+[RFDAP-XXXX]
```
Comment (lvoliveira) [inline .github/PULL_REQUEST_TEMPLATE.md]: Same here.
```suggestion
```

File: pyproject.toml
Comment (lvoliveira) [inline pyproject.toml:1]: Update the python versions here if you remove the 3.8 and 3.9

**Review by leogreal — APPROVED** [review]
Great work! 💯

File: pyproject.toml
Comment (ricardobz) [reply pyproject.toml:1]: Done!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 15 above)
Comment (ricardobz) [reply .github/PULL_REQUEST_TEMPLATE.md]: Done!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 14 above)
Comment (ricardobz) [reply .github/PULL_REQUEST_TEMPLATE.md]: Done!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 13 above)
Comment (ricardobz) [reply .github/PULL_REQUEST_TEMPLATE.md]: Done!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 12 above)
Comment (ricardobz) [reply .github/PULL_REQUEST_TEMPLATE.md]: Done!

File: .github/workflows/main.yml (hunk 11 above)
Comment (ricardobz) [reply .github/workflows/main.yml]: Removed!

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+<!--- Link if a related PR is open in another repository -->
+
+### How to test this feature/fix?
+
+<!--- Steps to verify the change works, including any edge cases worth checking -->
+
+### Additional information (screenshots, references, links):
+
+<!--- Replace XXXX with the issue code -->
```
Comment (lvoliveira) [inline .github/PULL_REQUEST_TEMPLATE.md]: Nit: you missed this one 🙂
```suggestion
```

File: tox.ini [hunk 17]
```diff
@@ -1,22 +1,20 @@
 [tox]
-envlist = py{36,37,38,39,310}-sqla{13,14}, linting, docs
+envlist = py{38,39,310,311,312}-sqla{14,20}, linting, docs
```
Comment (lvoliveira) [inline tox.ini]:
```suggestion
envlist = py{310,311,312}-sqla{14,20}, linting, docs
```

**Review by lvoliveira — APPROVED** [review]
Leaving just two last comments

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 16 above)
Comment (ricardobz) [reply .github/PULL_REQUEST_TEMPLATE.md]: Done!

File: tox.ini (hunk 17 above)
Comment (ricardobz) [reply tox.ini]: Done! Nice catch!
